test_provider.py

```diff
@@ -0,0 +1,454 @@
+        MODE = 'programMode'
+        ToO = 'tooType'
+        NOTES = 'INFO_SCHEDNOTE'
+
+    class _TAKeys(Enum):
+        CATEGORIES = 'timeAccountAllocationCategories'
+        CATEGORY = 'category'
+        AWARDED_TIME = 'awardedTime'
+        PROGRAM_TIME = 'programTime'
+        PARTNER_TIME = 'partnerTime'
+    
+    class _GroupsKeys(Enum):
+        KEY = 'GROUP_GROUP_SCHEDULING'
+        ORGANIZATIONAL_FOLDER = 'ORGANIZATIONAL_FOLDER'
+    
+    class _ObsKeys(Enum):
+        KEY = 'OBSERVATION_BASIC'
+        ID = 'observationId'
+        INTERNAL_ID = 'key'
+        QASTATE = 'qaState'
+        LOG = 'obsLog'
+        STATUS = 'obsStatus'
+        PRIORITY = 'priority'
+        TITLE = 'title'
+        SEQUENCE = 'sequence'
+        SETUPTIME_TYPE = 'setupTimeType'
+        SETUPTIME = 'setupTime'
+        OBS_CLASS = 'obsClass'
+        PHASE2 = 'phase2Status'
+
+    class _TargetKeys(Enum):
+        KEY = 'TELESCOPE_TARGETENV'
+        BASE = 'base'
+        TYPE = 'type'
+        RA = 'ra'
+        DEC = 'dec'
+        DELTARA = 'deltara'
+        DELTADEC = 'deltadec'
+        EPOCH = 'epoch'
+        TAG = 'tag'
+        MAGNITUDES = 'magnitudes'
+        NAME = 'name'
+
+    class _ConstraintsKeys(Enum):
+        KEY = 'SCHEDULING_CONDITIONS'
+        CC = 'cc'
+        IQ = 'iq'
+        SB = 'sb'
+        WV = 'wv'
+        ELEVATION_TYPE = 'elevationConstraintType'
+        ELEVATION_MIN = 'elevationConstraintMin'
+        ELEVATION_MAX = 'elevationConstraintMax'
+    
+    class _AtomKeys(Enum):
+        OBS_CLASS = 'observe:class'
+        INSTRUMENT = 'instrument:instrument'
+        WAVELENGTH = 'instrument:observingWavelength'
+        OBSERVED = 'metadata:complete'
+        TOTAL_TIME = 'totalTime'
+        OFFSET_P = 'telescope:p'
+        OFFSET_Q = 'telescope:q'
+    
+    class _TimingWindowsKeys(Enum):
+        TIMING_WINDOWS = 'timingWindows'
+        START = 'start'
+        DURATION = 'duration'
+        REPEAT = 'repeat'
+    
+    class _MagnitudeKeys(Enum):
+        NAME = 'name'
+        VALUE = 'value'
+
+    def __init__(self, path):
+        self.path = path
+
+    def load_program(self, path: str) -> dict:
+        with open(path, 'r') as f:
+            return json.loads(f.read())
+    
+    @staticmethod
+    def sort_observations(obs: List[Observation]) -> List[Observation]:
+        return sorted(obs, key=lambda x: x.order)
+    
+    @staticmethod
+    def strip_constraints(constraint: str):
+        return constraint.split('/')[0].split('%')[0]
+
+    @staticmethod
+    def parse_magnitude(json: dict) -> Magnitude:
+        band = MagnitudeBands[json['name']]
+        value = json['value']
+        error = None
+        return Magnitude(band, value, error)
+    
+    @staticmethod
+    def get_program_dates(program_type: ProgramTypes, id: str, notes: list) -> Tuple[datetime, datetime]:
+        start = end = None
+        year = int(id[3:7])
+        next_year = year + 1
+        semester = id[7]
```

Comment:
```
semester = SemesterHalf[id[7]]
```
Then we can work with `SemesterHalf.A` and `SemesterHalf.B` instead of `str` as we do up ahead.

Let's try to minimize our actual work with strings and push it as far back as it can go, so we are working with mini-model objects.